kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h

```diff
@@ -0,0 +1,714 @@
+        (pCurrentCondition)->EquationIdVector(rEquationId, rCurrentProcessInfo);
+
+        if (mUpdateAcceleration)
+        {
+            (pCurrentCondition)->CalculateMassMatrix(mMassMatrix[k], rCurrentProcessInfo);
+            AddDynamicsToRHS(pCurrentCondition, rRHS_Contribution,
+                             mDampingMatrix[k], mMassMatrix[k], rCurrentProcessInfo);
+        }
+        AddDynamicsToLHS(rLHS_Contribution, mDampingMatrix[k], mMassMatrix[k],
+                         rCurrentProcessInfo);
+
+        KRATOS_CATCH("")
+    }
+
+    void Condition_Calculate_RHS_Contribution(Condition::Pointer pCurrentCondition,
+                                              LocalSystemVectorType& rRHS_Contribution,
+                                              Element::EquationIdVectorType& rEquationId,
+                                              ProcessInfo& rCurrentProcessInfo) override
+    {
+        KRATOS_TRY;
+
+        int k = OpenMPUtils::ThisThread();
+
+        (pCurrentCondition)->InitializeNonLinearIteration(rCurrentProcessInfo);
+        (pCurrentCondition)->CalculateRightHandSide(rRHS_Contribution, rCurrentProcessInfo);
+        (pCurrentCondition)->CalculateLocalVelocityContribution(mDampingMatrix[k], rRHS_Contribution, rCurrentProcessInfo);
+        (pCurrentCondition)->EquationIdVector(rEquationId, rCurrentProcessInfo);
+
+        // adding the dynamic contributions (static is already included)
+        if (mUpdateAcceleration)
+        {
+            (pCurrentCondition)->CalculateMassMatrix(mMassMatrix[k], rCurrentProcessInfo);
+            AddDynamicsToRHS(pCurrentCondition, rRHS_Contribution,
+                             mDampingMatrix[k], mMassMatrix[k], rCurrentProcessInfo);
+        }
+        KRATOS_CATCH("");
+    }
+
+    void Clear() override
+    {
+        this->mpDofUpdater->Clear();
+    }
+
+    ///@}
+    ///@name Access
+    ///@{
+
+    ///@}
+    ///@name Inquiry
+    ///@{
+
+    ///@}
+    ///@name Input and output
+    ///@{
+
+    /// Turn back information as a string.
+    std::string Info() const override
+    {
+        return "ResidualBasedBossakVelocityScheme";
+    }
+
+    /// Print information about this object.
+    void PrintInfo(std::ostream& rOStream) const override
+    {
+        rOStream << Info();
+    }
+
+    /// Print object's data.
+    void PrintData(std::ostream& rOStream) const override
+    {
+        rOStream << Info();
+    }
+    ///@}
+    ///@name Friends
+    ///@{
+
+    ///@}
+
+protected:
+    ///@name Protected static Member Variables
+    ///@{
+
+    struct BossakConstants
+    {
+        double Alpha;
+        double Gamma;
+        double Beta;
+        double C0;
+        double C1;
+        double C2;
+        double C3;
+        double C4;
+        double C5;
+        double C6;
+    };
+
+    ///@}
+    ///@name Protected member Variables
+    ///@{
+
+    std::vector<LocalSystemVectorType> mSecondDerivativeValuesVectorOld;
```

Comment:
just to tell that the c++ way would be to use "thread_local" storage (not sure if it works on all compilers though)